diff --git a/craftos2-lua/src/lapi.c b/craftos2-lua/src/lapi.c
index df0ca5b..9609796 100644
--- a/craftos2-lua/src/lapi.c
+++ b/craftos2-lua/src/lapi.c
@@ -612,7 +612,7 @@ LUA_API const char *lua_pushsubstring (lua_State *L, int idx, size_t start, size
   ss->tsr.marked = luaC_white(g);
   ss->tsr.tt = LUA_TSUBSTR;
   ss->tsr.next = g->allgc;
-  g->allgc = ss;
+  g->allgc = (GCObject*)ss;
   ss->tss.cluster = cluster;
   ss->tss.str = str;
   ss->tss.offset = start - 1;
diff --git a/craftos2-lua/src/lgc.c b/craftos2-lua/src/lgc.c
index 4b0592d..0cd13ab 100644
--- a/craftos2-lua/src/lgc.c
+++ b/craftos2-lua/src/lgc.c
@@ -578,13 +578,13 @@ static void propagatemark (global_State *g) {
       break;
     }
     case LUA_TROPSTR: {
-      TString *tr = gco2tr(o);
+      TString *tr = (TString*)gco2tr(o);
       g->gray = tr->tsr.gclist;  /* remove from 'gray' list */
       size = traverserope(g, tr);
       break;
     }
     case LUA_TSUBSTR: {
-      TString *ss = gco2ss(o);
+      TString *ss = (TString*)gco2ss(o);
       g->gray = ss->tss.gclist;  /* remove from 'gray' list */
       size = traversesubstr(g, ss);
       break;
@@ -729,8 +729,8 @@ static void freeobj (lua_State *L, GCObject *o) {
       luaM_freemem(L, o, sizestring(gco2ts(o)));
       break;
     }
-    case LUA_TROPSTR: luaS_freerope(L, gco2tr(o)); break;
-    case LUA_TSUBSTR: luaS_freesubstr(L, gco2ss(o)); break;
+    case LUA_TROPSTR: luaS_freerope(L, (TString *)gco2tr(o)); break;
+    case LUA_TSUBSTR: luaS_freesubstr(L, (TString *)gco2ss(o)); break;
     default: lua_assert(0);
   }
 }
diff --git a/craftos2-lua/src/lstring.c b/craftos2-lua/src/lstring.c
index d124e79..e636b81 100644
--- a/craftos2-lua/src/lstring.c
+++ b/craftos2-lua/src/lstring.c
@@ -222,7 +222,7 @@ TString *luaS_concat (lua_State *L, TString *l, TString *r) {
   rope->tsr.marked = luaC_white(g);
   rope->tsr.tt = LUA_TROPSTR;
   rope->tsr.next = g->allgc;
-  g->allgc = rope;
+  g->allgc = (GCObject *)rope;
   rope->tsr.cluster = cluster;
   rope->tsr.left = l;
   rope->tsr.right = r;
